File: app/src/main/java/com/pengxh/autodingding/service/NotificationMonitorService.java
package com.pengxh.autodingding.service;

import android.app.Notification;
import android.content.ComponentName;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.service.notification.NotificationListenerService;
import android.service.notification.StatusBarNotification;
import android.text.TextUtils;
import android.util.Log;

import com.pengxh.androidx.lite.utils.SaveKeyValues;
import com.pengxh.androidx.lite.utils.TimeOrDateUtil;
import com.pengxh.autodingding.BaseApplication;
import com.pengxh.autodingding.bean.HistoryRecordBean;
import com.pengxh.autodingding.bean.MailInfo;
import com.pengxh.autodingding.bean.NotificationBean;
import com.pengxh.autodingding.greendao.HistoryRecordBeanDao;
import com.pengxh.autodingding.greendao.NotificationBeanDao;
import com.pengxh.autodingding.ui.WelcomeActivity;
import com.pengxh.autodingding.utils.Constant;
import com.pengxh.autodingding.utils.MailInfoUtil;
import com.pengxh.autodingding.utils.MailSender;

import java.util.UUID;

/**
 * @description: 状态栏监听服务
 * @author: Pengxh
 * @email: 290677893@qq.com
 * @date: 2019/12/25 23:17
 */
public class NotificationMonitorService extends NotificationListenerService {

    private static final String TAG = "MonitorService";
    private HistoryRecordBeanDao recordBeanDao;
    private NotificationBeanDao notificationBeanDao;

    /**
     * 有可用的并且和通知管理器连接成功时回调
     */
    @Override
    public void onListenerConnected() {
        Log.d(TAG, "onListenerConnected");
        recordBeanDao = BaseApplication.getDaoSession().getHistoryRecordBeanDao();
        notificationBeanDao = BaseApplication.getDaoSession().getNotificationBeanDao();
    }

    /**
     * 当有新通知到来时会回调
     */
    @Override
    public void onNotificationPosted(StatusBarNotification sbn) {
        Bundle extras = sbn.getNotification().extras;
        // 获取接收消息APP的包名
        String packageName = sbn.getPackageName();
        // 获取接收消息的内容
        String notificationText = extras.getString(Notification.EXTRA_TEXT);

        //保存所有通知信息
        NotificationBean notificationBean = new NotificationBean();
        notificationBean.setUuid(UUID.randomUUID().toString());
        notificationBean.setPackageName(packageName);
        notificationBean.setNotificationTitle(extras.getString(Notification.EXTRA_TITLE));
        notificationBean.setNotificationMsg(notificationText);
        notificationBean.setPostTime(TimeOrDateUtil.timestampToCompleteDate(System.currentTimeMillis()));
        notificationBeanDao.save(notificationBean);

//        if (packageName.equals("com.tencent.mobileqq")) {
        if (packageName.equals("com.alibaba.android.rimet")) {
            if (notificationText == null || notificationText.equals("")) {
                return;
            }
            Log.d(TAG, "onNotificationPosted ===> " + notificationText);
            if (notificationText.contains("考勤打卡")) {
                //保存打卡记录
                HistoryRecordBean bean = new HistoryRecordBean();
                bean.setUuid(UUID.randomUUID().toString());
                bean.setDate(TimeOrDateUtil.timestampToCompleteDate(System.currentTimeMillis()));
                bean.setMessage(notificationText);
                recordBeanDao.save(bean);

                String emailAddress = (String) SaveKeyValues.getValue(Constant.EMAIL_ADDRESS, "");
                if (TextUtils.isEmpty(emailAddress)) {
                    Log.d(TAG, "邮箱地址为空");
                } else {
                    //发送打卡成功的邮件
                    new Thread(() -> {
                        MailInfo mailInfo = MailInfoUtil.createMail(emailAddress, notificationText);
                        MailSender.getSender().sendTextMail(mailInfo);
                    }).start();

                    Intent intent = new Intent(this, WelcomeActivity.class);
                    intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                    startActivity(intent);
                }
            }
        }
    }

    /**
     * 当有通知移除时会回调
     */
    @Override
    public void onNotificationRemoved(StatusBarNotification sbn) {

    }

    @Override
    public void onListenerDisconnected() {
        Log.d(TAG, "onListenerDisconnected");
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            // 通知侦听器断开连接 - 请求重新绑定
            requestRebind(new ComponentName(this, NotificationListenerService.class));
        }
    }
}

File: app/src/main/java/com/pengxh/autodingding/service/NotificationMonitorService.kt
package com.pengxh.autodingding.service

import android.app.Notification
import android.os.BatteryManager
import android.service.notification.NotificationListenerService
import android.service.notification.StatusBarNotification
import android.util.Log
import com.pengxh.autodingding.BaseApplication
import com.pengxh.autodingding.bean.NotificationBean
import com.pengxh.autodingding.extensions.backToMainActivity
import com.pengxh.autodingding.extensions.createTextMail
import com.pengxh.autodingding.extensions.openApplication
import com.pengxh.autodingding.extensions.sendTextMail
import com.pengxh.autodingding.fragment.SettingsFragment
import com.pengxh.autodingding.utils.Constant
import com.pengxh.kt.lite.extensions.getSystemService
import com.pengxh.kt.lite.extensions.show
import com.pengxh.kt.lite.extensions.timestampToCompleteDate
import com.pengxh.kt.lite.utils.SaveKeyValues
import java.util.UUID

/**
 * @description: 状态栏监听服务
 * @author: Pengxh
 * @email: 290677893@qq.com
 * @date: 2019/12/25 23:17
 */
class NotificationMonitorService : NotificationListenerService() {

    private val kTag = "MonitorService"

    private val notificationBeanDao by lazy { BaseApplication.get().daoSession.notificationBeanDao }
    private val batteryManager by lazy { getSystemService<BatteryManager>() }

    /**
     * 有可用的并且和通知管理器连接成功时回调
     */
    override fun onListenerConnected() {
        Log.d(kTag, "onListenerConnected: 通知监听服务运行中")
        SettingsFragment.weakReferenceHandler?.sendEmptyMessage(Constant.NOTICE_LISTENER_CONNECTED_CODE)
    }

    /**
     * 当有新通知到来时会回调
     */
    override fun onNotificationPosted(sbn: StatusBarNotification) {
        val extras = sbn.notification.extras
        // 获取接收消息APP的包名
        val packageName = sbn.packageName
        // 获取接收消息的标题
        val title = extras.getString(Notification.EXTRA_TITLE) ?: ""
        // 获取接收消息的内容
        val notice = extras.getString(Notification.EXTRA_TEXT)
        if (notice.isNullOrBlank()) {
            return
        }
        SettingsFragment.weakReferenceHandler?.sendEmptyMessage(Constant.NOTICE_LISTENER_CONNECTED_CODE)

        if (notice != Constant.FOREGROUND_RUNNING_SERVICE_TITLE) {
            val notificationBean = NotificationBean()
            notificationBean.uuid = UUID.randomUUID().toString()
            notificationBean.packageName = packageName
            notificationBean.notificationTitle = title
            notificationBean.notificationMsg = notice
            notificationBean.postTime = System.currentTimeMillis().timestampToCompleteDate()
            notificationBeanDao.save(notificationBean)
        }

        val emailAddress = SaveKeyValues.getValue(Constant.EMAIL_ADDRESS, "") as String
        if (emailAddress.isEmpty()) {
            "邮箱地址为空".show(this)
            return
        }

        if (packageName == Constant.DING_DING) {
            if (notice.contains("成功")) {
                backToMainActivity()
                "即将发送通知邮件，请注意查收".show(this)
                val subject = SaveKeyValues.getValue(
                    Constant.EMAIL_TITLE, "打卡结果通知"
                ) as String
                notice.createTextMail(subject, emailAddress).sendTextMail()
            }
        } else if (packageName == Constant.WECHAT || packageName == Constant.QQ || packageName == Constant.TIM || packageName == Constant.ZFB) {
            if (notice.contains("电量")) {
                val capacity = batteryManager?.getIntProperty(
                    BatteryManager.BATTERY_PROPERTY_CAPACITY
                )
                "当前手机剩余电量为：${capacity}%".createTextMail(
                    "查询手机电量通知", emailAddress
                ).sendTextMail()
            } else {
                val key = SaveKeyValues.getValue(Constant.DING_DING_KEY, "打卡") as String
                if (notice.contains(key)) {
                    openApplication(Constant.DING_DING, true)
                }
            }
        }
    }

    /**
     * 当有通知移除时会回调
     */
    override fun onNotificationRemoved(sbn: StatusBarNotification) {}

    override fun onListenerDisconnected() {
        Log.d(kTag, "onListenerDisconnected: 通知监听服务已关闭")
        SettingsFragment.weakReferenceHandler?.sendEmptyMessage(Constant.NOTICE_LISTENER_DISCONNECTED_CODE)
    }
}